File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    int splitArray(vector<int>& nums, int m) {
         int n = nums.size();
        vector<vector<int>>dp(n+1, vector<int>(m+1, INT_MAX));
        dp[0][0] = 0;
        for(int i = 1; i <= n; i++){
            for(int j = 1; j<=m; j++){
                int cur = 0;
                for(int k = i; k > 0; k--){
                    cur += nums[k-1];
                    dp[i][j] = min(dp[i][j], max(cur, dp[k-1][j-1]));
                }
            }
        }
        return dp[n][m];
    }
};

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {
    public int splitArray(int[] nums, int k) {
        if (nums == null || nums.length == 0 || k == 0 ) {
            return 0;
        }
        int maxi = Integer.MIN_VALUE;
        int sum=0;
         int ans=0;
        for(int i=0;i<nums.length;i++){
            sum += nums[i];
            maxi = Math.max(maxi,nums[i]);
        }
        if(k == 1)return sum;
        else if(k == nums.length)return maxi;
        else{
           
            int l = maxi;
            int r = sum;
            while(l <= r){
                int m = l + (r-l)/2;
                if(isPossible(nums,k, m)){
                    ans=m;
                    r = m-1;
                }else{
                    l = m + 1;
                }
            }
        }
        return ans;
    }
    static boolean isPossible(int [] nums,int k,int m){
        int sum=0;
        int partitions =1;
        for(int i=0;i<nums.length;i++){
            sum += nums[i];
            if(sum > m ){
                partitions++;
                sum = nums[i];
            }
        }
        if(partitions <= k)return true;
        return false;
    }
}